arch/arm/core/cortex_m/mpu/Kconfig

```diff
@@ -62,6 +62,31 @@ config MPU_ALLOW_FLASH_WRITE
 	help
 	  Enable this to allow MPU RWX access to flash memory
 
+config MPU_SRAM_GAP_FILLING
```

Comment:
@ulfalizer can we put the definition into `if USERSPACE`  block, to make it defined only if USERSPACE is also defined? Then, what will happen with the dependency on `MPU_REQUIRES_NON_OVERLAPPING_REGIONS`?

Reply:
(Note that `if FOO` is just a shorthand for adding `depends on FOO` to the symbol(s) within it. `if` and `depends on` are exactly the same thing. I think this is something that confuses people re. Kconfig, where people assume `if` does more than that. :)

If a symbol is defined in multiple locations and should depend on `USERSPACE`, then all definition locations need to have `depends on USERSPACE` (or be within an `if USERSPACE` -- same thing), since the dependencies get ORed together.

Personally, I'd try to refactor it so that there's just a single definition in `arch/Kconfig`, maybe along with some helper symbols.

Reply:
>Personally, I'd try to refactor it so that there's just a single definition in arch/Kconfig, maybe along with some helper symbols.

That's hard; I raised the definition up in arch/Kconfig, as it is essentially a cross-arch option, but I want to keep re-defining it in arch/arm/cortex-m/mpu/Kconfig , so I can have the additional help text that is ARMv8-M specific. But it is becoming clear from your comments that dependencies are always ORed so I cannot add additional dependencies here (so I need to move depends on USERSPACE up in the top-level definition, correct?)

So how can I do it in a way it satisfies you @ulfalizer ? :)

Reply:
These two are 100% identical:

```
if FOO
if BAR

config FOO
	bool "foo"
	depends on BAZ

endif
endif
```

```
config FOO
	bool "foo"
	depends on FOO && BAR && BAZ
```

`if` does not "cancel" the definition, but only modified the dependencies, like `depends on` does. When a symbol is defined more than once, the dependencies get `||`'d together.

Reply:
A tip is to play around with this in `menuconfig` btw, because it'll make it clear how multiple definitions work:

```
config DEP1
	bool "dep1"

config DEP2
	bool "dep2"

config MULTIDEF
	bool "first multidef prompt"
	depends on DEP1

config MULTIDEF
	bool "second multidef prompt"
	depends on DEP2
```

Reply:
> These two are 100% identical:
> 
> ```
> if FOO
> if BAR
> 
> config FOO
> 	bool "foo"
> 	depends on BAZ
> 
> endif
> endif
> ```
> 
> ```
> config FOO
> 	bool "foo"
> 	depends on FOO && BAR && BAZ
> ```
> 
> `if` does not "cancel" the definition, but only modified the dependencies, like `depends on` does. When a symbol is defined more than once, the dependencies get `||`'d together.

Understood. So can you help me finding a solution that satisfies my needs?
- I want MPU_SRAM_GAP_FILLING to be defined high in arch/Kconfig, as it may be used by multiple ARCHes
- I want it to depend on MPU_REQUIRES_NON_OVERLAPPING_REGIONS **and** USERSPACE
- But: I want the help text to be somewhat ARM-specific, that's why I want to have the option re-defined in the ./arm sub-directory. 

So you suggest keeping the solution **as is** and simply add the same list of dependencies in the two definitions of the symbol?

Reply:
> but I want to keep re-defining it in arch/arm/cortex-m/mpu/Kconfig, so I can have the additional help text that is ARMv8-M specific

I'd have a single definition and some ARM-specific notes in it (well sectioned off). Think it's less of a mess than dealing with multiple definition locations.

Otherwise, the only option is to add `depends on USERSPACE` to all definitions, if the symbol should (always) depend on `USERSPACE`.

Reply:
Another option would be to have multiple definitions, but only putting them in `arch`-specific Kconfig files, and not in the generic `arch/Kconfig`.

Reply:
> So you suggest keeping the solution **as is** and simply add the same list of dependencies in the two definitions of the symbol?

Yep, would have to do that, if the definitions are kept as-is.

Reply:
@ulfalizer thanks. There's also an option to have MPU_SRAM_GAP_FILLING defined in arch/Kconfig,
```
config MPU_SRAM_GAP_FILLING
    bool "bla bla bla, to make it visible "
    depends on MPU_REQUIRES_NON_OVERLAPPING_REGIONS
    help
       Some generic cross-ARCH text
```

 and have an arm-specific option in arm/cortex-m/mpu/Kconfig as:
```
config ARM_MPU_SRAM_GAP_FILLING
    bool
    depends on USERSPACE
    depends on MPU_SRAM_GAP_FILLING
    default y
    help 
      Some ARM-specific text
```

Reply:
How about the last option??

>Another option would be to have multiple definitions, but only putting them in arch-specific Kconfig files, and not in the generic arch/Kconfig.

We thought of not doing this, as it is essentially the same functionality used in different ARCHes

Reply:
Yeah, that might work too. One drawback is that it'll still be possible to enable `MPU_SRAM_GAP_FILLING` when `USERSPACE` is `n`. Would that be confusing?

Reply:
No. If someone does it, it will simply be ignored by the underlying, ARM-specific option, which adds the USERSPACE dependency, and that's totally OK.

Reply:
The point is that : only the ARCH-specific option will be used in the code. the top-arch option will only be used in Kconfig, to enable the ARM-specific option.

Reply:
Was thinking from an interface design perspective. There'll be a button you toggle that says "MPU SRAM gap filling" or the like, but it won't do anything unless you also toggle "User mode threads".

Feels like it gets lost sometimes that Kconfig defines a configuration interface. :)

Reply:
As long as that wouldn't be weird to users, it'll work at least.

Reply:
My preferred solution would be either

1) A single definition in `arch/Kconfig`, with some arch-specific notes in the help text, and maybe some helper symbols to adjust it for different arch capabilities

2) Have multiple arch-specific definitions, and skip the definition in `arch/Kconfig`

I won't impose anything though. Just keep the multiple-definition dependency gotchas in mind.